src/python/pants/backend/helm/goals/package.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+import os
+from dataclasses import dataclass
+
+from pants.backend.helm.target_types import HelmChartFieldSet, HelmChartOutputPathField
+from pants.backend.helm.util_rules.chart import HelmChart, HelmChartMetadata, HelmChartRequest
+from pants.backend.helm.util_rules.tool import HelmProcess
+from pants.core.goals.package import BuiltPackage, BuiltPackageArtifact, PackageFieldSet
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, Directory, MergeDigests, RemovePrefix
+from pants.engine.process import ProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class BuiltHelmArtifact(BuiltPackageArtifact):
+    name: str | None = None
+    metadata: HelmChartMetadata | None = None
+
+    @classmethod
+    def create(cls, path_to_dir: str, chart_metadata: HelmChartMetadata) -> BuiltHelmArtifact:
+        path = os.path.join(path_to_dir, _helm_artifact_filename(chart_metadata))
+        return cls(
+            name=chart_metadata.artifact_name,
+            metadata=chart_metadata,
+            relpath=path,
+            extra_log_lines=(f"Built Helm chart artifact: {path}",),
+        )
+
+
+def _helm_artifact_filename(chart_metadata: HelmChartMetadata) -> str:
+    return f"{chart_metadata.artifact_name}.tgz"
+
+
+@dataclass(frozen=True)
+class HelmPackageFieldSet(HelmChartFieldSet, PackageFieldSet):
+    output_path: HelmChartOutputPathField
+
+
+@rule(desc="Package Helm chart", level=LogLevel.DEBUG)
+async def run_helm_package(field_set: HelmPackageFieldSet) -> BuiltPackage:
+    result_dir = "__out"
+
+    chart, result_digest = await MultiGet(
+        Get(HelmChart, HelmChartRequest(field_set)),
+        Get(Digest, CreateDigest([Directory(result_dir)])),
+    )
+
+    input_digest = await Get(Digest, MergeDigests([chart.snapshot.digest, result_digest]))
+    process_output_file = os.path.join(result_dir, _helm_artifact_filename(chart.metadata))
+
+    process_result = await Get(
+        ProcessResult,
+        HelmProcess(
+            argv=["package", chart.path, "-d", result_dir],
+            input_digest=input_digest,
+            output_files=(process_output_file,),
+            description=f"Packaging Helm chart: {field_set.address.spec_path}",
+        ),
+    )
+
+    stripped_output_digest = await Get(
+        Digest, RemovePrefix(process_result.output_digest, result_dir)
+    )
+
+    artifact_path = field_set.output_path.value_or_default(file_ending=None)
```

Comment:
Due to Helm convention of producing chart packages that follow the pattern `[chart_name]-[chart_version].tgz`, the `output_path` used here is not interpreted as a file name (as other `package` goal implementations do), but rather as a directory name.

This can be confusing to users as it is a _special case_ for the `output_path` field in this target compared to other targets. The implementation of the special case can be observed in the `HelmChartOutputPathField` implementation:

```python
def value_or_default(self, *, file_ending: str | None) -> str:
      if self.value:
          return self.value
      return os.path.join(self.address.spec_path.replace(os.sep, "."))
```

This special case could be removed by slightly modifying the signature of the `valur_or_default` method in the base `OutputPathField` class so it can take an optional `file_suffix` argument (which in this case would represent the chart version). With this modification, the `value_or_default` method would be used as such:

```python
artifact_path = field_set.output_path.value_or_default(file_suffix=chart.metadata.version, file_ending="tgz")
```

An alternative to this could be not consider file suffixes at all and use what `output_path` tells us as the final file name. In this case when the `publish` goal gets implemented, then a file rename would need to be done before the actual upload of the package artifact as the previously mentioned artifact name format is a Helm strict requirement as mentioned here:
https://helm.sh/docs/topics/registries/

Reply:
Directory-based is how PyOxidizer works too :) 

https://github.com/pantsbuild/pants/blob/f4d6b52beb16725b33f07e69f2066ae2657f7b84/src/python/pants/backend/python/packaging/pyoxidizer/target_types.py#L19-L36

Feel free to take any of this language to expand on the `help` for Helm target.